Updated Style Guide (#596)
saucelabs/sauce-docs: docs/contributing/style-guide.md

---
id: style-guide
title: Style Guide
sidebar_label: Style Guide
description: The Sauce Labs Documentation Style Guide
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

Sauce Labs welcomes your contributions to our documentation!
This document describes the options available for creating content for the site,
along with some guidelines and conventions.

## Markdown

This site uses Docusaurus version 2 beta.The 
[Docusaurus Markdown Features documentation](https://docusaurus.io/docs/markdown-features) 
is interesting, but is missing descriptions of several key features. Markdown Guide provides a more comprehensive description of available 
[Basic Syntax](https://www.markdownguide.org/basic-syntax) and 
[Extended Syntax](https://www.markdownguide.org/extended-syntax/) with some helpful tips for how to use it well.
Markdown is used for its readability, but for anything where you need more control or more complex styling,

> plain HTML may also be used directly in the file.

Author: @spider-sauce do you know what extended markdown isn't supported by github markdown? You had linked to the GHM spec (which is the most correct), but it's just not helpful at all for someone trying to figure out how to write markdown. :)

> plain HTML may also be used directly in the file.

Author: nevermind, found the details.

## Frontmatter

At the top of each docs page, you need to include these things: 

| Variable | Description |
| ----------- | ----------- |

> |id|A brief string that uniquely identifies the page within its parent folder. The id and the name of the file should be the same.|
> |title|The main title of the page. This value will automatically be rendered using the H1 style at the top of the page.
> |sidebar_label|This is what will appear in the left hand navigation tree for the page.|
> |description (optional)|This is what appears when the page is referenced in a Google search result.|

Reviewer: ```suggestion
|`id`|A brief string that uniquely identifies the page within its parent folder. The id and the name of the file should be the same.|
|`title`|The main title of the page. This value will automatically be rendered using the H1 style at the top of the page.
|`sidebar_label`|This is what will appear in the left hand navigation tree for the page.|
|`description` (optional)|This is what appears when the page is referenced in a Google search result.|
|`tags` (optional)| A list of terms that help categorize the page for SEO purposes.|
```

> ### Code Example

Reviewer: I'm not sure this needs to be a heading? It feels like overkill for the floating TOC. Thoughts, @spider-sauce ?

Author: I was trying to break up the H2 sections somehow. Might be a better option to use a horizonal line (`---`)

Reviewer: Not sure why you want to "break up the H2 sections?" Do you think it makes the floating tOC hard to read when it's just a list of H2s? Or is it in the main body of the page that you object to it? It might be helpful to look at the formatting of a CLI or the saucectl config pages to see how we handle property definitions that include syntax samples: https://docs.saucelabs.com/testrunner-toolkit/saucectl

Author: Oh, yes, this it totally the right way to do it, I'll fix

Author: I tried, this and didn't get it to look good, so I went with putting everything into a table. Either way the H3s are removed.

```markdown
---
id: style-guide
title: Style Guide Introduction
sidebar_label: Style Guide
description: The Sauce Labs Documentation Style Guide
---
```

## Introduction

The first paragraph of the documentation should set the user's expectations for what they will find on the page. 
Describe the key benefits to the user, but do not include links.

## Headers

For accessibility and SEO reasons, never have an H4 header that isn't under an H3 header, 
or an H3 header that isn't under an H2 header.

* H1 headers should never be used in a document since the title is automatically generated as an H1.

> * H2 headers are used for SEO, so make sure they succinctly represent what a user will find on the page in that section
> * H3 headers are included in the page's table of contents on the right, so make sure the title describes something
> a user might want to click on

Reviewer: ```suggestion
* H2 headers are used for SEO, so make sure they succinctly represent what a user will find on the page in that section.
* H3 headers are included in the page's table of contents on the right, so make sure the title describes something
a user might want to click on.
```

* H4 headers are to emphasize things within a subsection of the page; these can be longer than the other headers if needed. 

### Code Example

```markdown
## H2 Header
### H3 Header
#### H4 Header
```

## Content

All words are rendered in the same paragraph even with line breaks, so long as there isn't an empty line.
As such, it is good practice for each line to be less than 120 characters long for readability, when possible.

### Code Example

```markdown
This
will
all
be
on
one
line.

The empty line above creates a new paragraph.

This forces a soft return<br />
rather than creating a new paragraph
```

### Display

This
will
all
be
on
one
line.

The empty line above creates a new paragraph.

This forces a soft return<br />

> rather than creating a new paragraph

Reviewer: It might be helpful to have these two sections in a table, so you can see the code and the rendered content side by side.

Author: Yeah, I couldn't do this side-by-side, but I got something working vertically.

## Images

All image files should be included in the `sauce-docs/static/img` directory, in a sub-directory that 
corresponds to the referencing directory. (e.g., images for a document in the `sauce-docs/docs/contributing` directory 
would be located in the `sauce-docs/static/img/contributing` directory.

To add an image from that directory, you need to import a special method by placing this line
below the [Frontmatter](#frontmatter), but above the [Introduction](#introduction)

```markdown
import useBaseUrl from '@docusaurus/useBaseUrl';
```

and then reference the image as follows:
```markdown
<img src={useBaseUrl('img/contributing/my-image.png')} alt="All images should have alt text" width="250"/>

> ```

Reviewer: This is correct, but this is actually HTML, not markdown, and is in conflict with the markdown image ref I commented on in the markdown-styles file of this review.

## Videos

Any referenced videos need to be from a Sauce Labs YouTube account.
The suggested iframe code structure is as follows:
```markdown
<iframe width="560" height="315" src="https://www.youtube.com/embed/-RDh1ukLN8w" frameborder="0" allow="accelerometer; 
autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```

## Tabs

> Tabs are a great option for when an example is different in different contexts. The primary usage of
> Tabs on this site is for code-specific descriptions or examples, since many such things on our platform
> are relevant to multiple languages.
> Ideally these examples will exist in Java, Node.js, Python, Ruby, C#, and the tabs should be placed in that order.

Reviewer: ```suggestion
Tabs are a great option when an example is different in different contexts. The primary usage of
tabs on this site is to illustrate the same example in multiple languages. Ideally, these examples will exist in Java, Node.js, Python, Ruby, and C#, and the tabs should be placed in that order.

:::tip
When a page includes multiple sets of tabs, use a `groupId` so when the user selects a particular tab, all the set on the page will switch to that tab.
:::
```

To use tabs, you need to import two special methods by placing these lines below the [Frontmatter](#frontmatter),
but above the [Introduction](#introduction):
```markdown
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
```

### Code Example:
```markdown

> <Tabs
>   defaultValue="java"

Reviewer: ```suggestion
<Tabs
  groupId="lang-ex"
  defaultValue="java"
```

Author: So... this is really cool and I didn't realize it, need to go update my other projects with it. Great tip.

to specify which language you are using, and it's a good idea to include a title with the code block as well.

### Example Code

    ```java title="Custom Title"
    // Java Code Example
    ```

### Display

```java title="Custom Title"
// Java Code Example
```

## Code References

The Sauce Labs Open Source Team created a plugin for use with Docusaurus to allow us to reference code on Github
rather than duplicating code in this repo. Ideally all code displayed in the Sauce Labs documentation points to code in
one of the `demo-<language>` repos on [Sauce Labs Training Github Org](https://github.com/saucelabs-training/?q=demo#org-repositories)
When referencing code, Include the language, "reference" and a title indicating what the code is.
The URL for the link can be for the entire file, or include specific line numbers at the end.
Additionally, the referenced code should reference a tag name rather than just a branch name (which can easily break) or
a commit hash (which is hard to version). For example, the `demo-java` repo has a tag for `website-1.0`, so the reference
link should specify this.

### Example Code

>     ```java reference title="Example Test"
>     https://github.com/saucelabs-training/demo-java/blob/website-1.0/selenium-examples/src/test/java/com/saucedemo/junit5/SauceBindingsTest.java#L35-L43
>     ```

Reviewer: ```suggestion
```java reference title="Example Test"
https://github.com/saucelabs-training/demo-java/blob/website-1.0/selenium-examples/src/test/java/com/saucedemo/junit5/SauceBindingsTest.java#L35-L43
```
```

Author: This one actually needs to be indented so that the tick marks display on the page, otherwise it treats it as a code block, which we don't want.

### Display

```java reference title="Example Test"
https://github.com/saucelabs-training/demo-java/blob/website-1.0/selenium-examples/src/test/java/com/saucedemo/junit5/SauceBindingsTest.java#L35-L43
```

## Admonitions

There are 4 types of [Docusaurus Admonitions](https://docusaurus.io/docs/markdown-features/admonitions):
* Note - Relevant information.
* Tip - A user should do this.
* Caution - A user should pay attention to this.

> * Warning - A user might do something dangerous!

Reviewer: There is actually a 5th one -- important -- but we haven't really publicized it and it doesn't have a custom color. It's just slightly darker gray than the note version. If we want to document it, we should probably create a style to differentiate it a little more.

|`description` (optional)|This is what appears when the page is referenced in a Google search result.|
|`tags` (optional)| A list of terms that help categorize the page for SEO purposes.|

It looks like this in the document:
```markdown
---
id: style-guide
title: Style Guide Introduction
sidebar_label: Style Guide
description: The Sauce Labs Documentation Style Guide
tags: [contributing, style, markdown]
---
```

## Introduction

The first paragraph of the documentation should set the user's expectations for what they will find on the page. 
Describe the key benefits to the user, but do not include links.

## Headers

For accessibility and SEO reasons, never have an H4 header that isn't under an H3 header, 
or an H3 header that isn't under an H2 header.

* H1 headers should never be used in a document since the title is automatically generated as an H1.
* H2 headers are used for SEO, so make sure they succinctly represent what a user will find on the page in that section.
* H3 headers are included in the page's table of contents on the right, so make sure the title describes something
a user might want to directly navigate to.

> * H4 headers are to emphasize things within a subsection of the page; these can be longer than the other headers if needed. 

Author: Oh yes, I like that. Maybe: 
> these can be longer than the other headers if needed because they aren't included in the Table of Contents

To add an image from that directory, you need to import a special method by placing this line
below the [Frontmatter](#frontmatter), but above the [Introduction](#introduction)

```markdown
import useBaseUrl from '@docusaurus/useBaseUrl';
```

and then reference the image as follows:
```html
<img src={useBaseUrl('img/contributing/my-image.png')} alt="All images should have alt text" width="250"/>
```

## Videos

Any referenced videos need to be from a Sauce Labs YouTube account.
The suggested iframe code structure is as follows:
```html
<iframe width="560" height="315" src="https://www.youtube.com/embed/-RDh1ukLN8w" frameborder="0" allow="accelerometer; 
autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```

## Tabs

Tabs are a great option when an example is different in different contexts. 
The primary usage of tabs on this site is to illustrate the same example in multiple languages. 
Ideally, these examples will exist in Java, Node.js, Python, Ruby, and C#, and the tabs should be placed in that order.

:::tip
When a page includes multiple sets of tabs, use a `groupId` so when the user selects a particular tab, 

> all the set on the page will switch to that tab.

Author: How about this instead:
> When a page includes multiple sets of tabs, use a `groupId` so when the user selects a particular tab, 
all tabs with that ID will switch to the selected tab.

> all the set on the page will switch to that tab.

Reviewer: Beautiful.

    ```

</td>
</tr>
<tr>
<td>

**Display**

</td>
      <td>

```java title="Custom Title"
RemoteWebDriver driver = new RemoteWebDriver(url, capabilities);
```

</td>
    </tr>
  </tbody>
</table>

## Code References

The Sauce Labs Open Source Team created a plugin for use with Docusaurus to allow us to reference code on Github
rather than duplicating code in this repo. Ideally all code displayed in the Sauce Labs documentation points to code in
one of the `demo-<language>` repos on [Sauce Labs Training Github Org](https://github.com/saucelabs-training/?q=demo#org-repositories)
When referencing code, Include the language, "reference" and a title indicating what the code is.
The URL for the link can be for the entire file, or include specific line numbers at the end.
Additionally, the referenced code should reference a tag name rather than just a branch name (which can easily break) or
a commit hash (which is hard to version). For example, the `demo-java` repo has a tag for `website-1.0`, so the reference
link should specify this.

<table class="code">
  <tbody>
    <tr>
      <td>

**Markdown**

</td>
      <td>

>     ```java reference title="Example Test"

Author: It's new! We have to update to the latest version of the plugin to get it to work. They hadn't even released it when I last did the PR, so I'll add the update to this PR.

* Note - Relevant information.
* Tip - A user should do this.
* Caution - A user should pay attention to this.
* Warning - A user might do something dangerous!

<table class="code">
  <tbody>
    <tr>
      <td>

**Markdown**

</td>
      <td>

```markdown
:::note
Relevant information for you.
:::

:::tip
You should do this.
:::

:::caution
You should probably pay attention to this.
:::

:::warning
You are about to do something dangerous!
:::
```

</td>
</tr>
<tr>
<td>

**Display**

</td>
      <td>

:::note
Relevant information for you.
:::

:::tip
You should do this.
:::

:::caution
You should probably pay attention to this.
:::

:::warning
You are about to do something dangerous!
:::

</td>
    </tr>
  </tbody>
</table>

## Badges

Badges are color-coded images that alert a reader to special considerations for a given section. 
We currently support 2 types of badges:

GOLD indicates that a page is deprecated or out-of-date in some way, for example:

* DEPRECATED

BLUE indicates that the feature being documented is limited in its scope or availability, for example:

* ENTERPRISE ONLY
* IOS ONLY
* LIVE TESTING ONLY
* BETA
* EARLY ACCESS

> 

Reviewer: I have never managed to add this to our internal guidelines, and our badges have been a bit of an evolving entity for some time, but often, when the badge (such as `REAL DEVICES ONLY` is in a table cell (such as a table of property definitions), we set it in the <small></small> tag so it is not so overwhelming.You can see an example of what I mean here: https://docs.saucelabs.com/testrunner-toolkit/configuration/cypress#tunnel 9Look at the `tunnel` property.

Author: I split the badges and highlights into H3's with examples

<table class="code">
  <tbody>
    <tr>
      <td>

**HTML**

</td>
      <td>

```html
<p><span className="sauceDBlue">BETA</span></p>
<p><span className="sauceDBlue">ENTERPRISE ONLY</span></p>
<p><span className="sauceDBlue">IOS ONLY</span></p>
<p><span className="sauceDBlue">LIVE TESTING ONLY</span></p>
<p><span className="sauceDBlue">EARLY ACCESS</span></p>

<p><span className="sauceGold">DEPRECATED</span></p>
```

</td>
</tr>
<tr>
<td>

**Display**

</td>
      <td>
<p><span className="sauceDBlue">BETA</span></p>
<p><span className="sauceDBlue">ENTERPRISE ONLY</span></p>
<p><span className="sauceDBlue">IOS ONLY</span></p>
<p><span className="sauceDBlue">LIVE TESTING ONLY</span></p>
<p><span className="sauceDBlue">EARLY ACCESS</span></p>

<p><span className="sauceGold">DEPRECATED</span></p>
</td>
    </tr>
  </tbody>
</table>

## Cards

> For overview pages that have 4 categories, we often use these Boxes. Note that you can't use Markdown inside this html.

Reviewer: Do we want to show the boxes that we have defined for a top or bottom row that is full width, or is that overkill? See this page for an example: https://docs.saucelabs.com/performance

Author: tbh, I almost didn't include this section because I doubt any potential external contributors will use these, and if Saucers want to use it, they'll just copy from another page that. :)
So, this is totally your call on whether it belongs as an example in the style guide.
